EventListener/TokenSubscriber.php

```diff
@@ -323,4 +338,137 @@ private function getCustomFieldValues(CustomObject $customObject, Token $token,
+        $fieldValues = [];
+
+        foreach ($customItems as $customItemData) {
+            // Name is known from the CI data array.
+            if ('name' === $customFieldAlias) {
+                $fieldValues[] = $customItemData['name'];
+
+                continue;
+            }
+
+            // Custom Field values are handled like this.
+            $customItem = new CustomItem($customObject);
+            $customItem->populateFromArray($customItemData);
+            $customItem = $this->customItemModel->populateCustomFields($customItem);
+
+            try {
+                $fieldValue = $customItem->findCustomFieldValueForFieldAlias($customFieldAlias);
+                // If the CO item doesn't have a value, get the default value
+                if (empty($fieldValue->getValue())) {
+                    $fieldValue->setValue($fieldValue->getCustomField()->getDefaultValue());
+                }
+                $fieldValues[] = $fieldValue->getCustomField()->getTypeObject()->valueToString($fieldValue);
+            } catch (NotFoundException $e) {
+                // Custom field not found.
+            }
+        }
+
+        return $this->tokenFormatter->format($fieldValues, TokenFormatter::DEFAULT_FORMAT);
+    }
+
+    /**
+     * @param array<mixed> $filters
+     *
+     * @return array<mixed>
+     */
+    private function getCustomFieldDataForLead(array $filters, int $leadId): array
+    {
+        $customFieldValues = [];
+        foreach ($filters as $condition) {
+            try {
+                if ('custom_object' !== $condition['object']) {
+                    continue;
+                }
+
+                if ('cmf_' === substr($condition['field'], 0, 4)) {
+                    $customField  = $this->customFieldModel->fetchEntity(
+                        (int) explode('cmf_', $condition['field'])[1]
+                    );
+                    $customObject = $customField->getCustomObject();
+                    $fieldAlias   = $customField->getAlias();
+                } elseif ('cmo_' === substr($condition['field'], 0, 4)) {
+                    $customObject = $this->customObjectModel->fetchEntity(
+                        (int) explode('cmo_', $condition['field'])[1]
+                    );
+                    $fieldAlias   = 'name';
+                } else {
+                    continue;
+                }
+
+                $result = $this->getCustomFieldValue($customObject, $leadId, $fieldAlias);
+
+                $customFieldValues[$condition['field']] = $result;
+            } catch (NotFoundException|InvalidCustomObjectFormatListException $e) {
+                continue;
+            }
+        }
+
+        return $customFieldValues;
+    }
+
+    /**
+     * @return array<mixed>
+     */
+    private function getCustomItems(CustomObject $customObject, int $leadId): array
+    {
+        $key = $customObject->getAlias().'-'.$leadId;
+        if (isset($this->customItems[$key])) {
```

Comment:
Do not use lazy loading here as we send emails to tens of thousands contacts within single script run. Remove the lazy loading and move executing of that query to another place, so that query is not executed twice for a single contact.